server/src/main/java/io/crate/planner/operators/HashJoin.java

```diff
@@ -122,9 +122,11 @@ public ExecutionPlan build(DependencyCarrier executor,
         LogicalPlan leftLogicalPlan = lhs;
         LogicalPlan rightLogicalPlan = rhs;
 
+        long lhsNumExpectedRows = lhs.numExpectedRows() < 0 ? 0 : lhs.numExpectedRows();
```

Comment:
It's out of the scope of this PR but maybe we can consider some new meta issue - check all usages of `numExpectedRows`, whether we need to handle -1 in all those places - sometimes we [multiply](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/planner/operators/GroupHashAggregate.java#L96) to this number.  Not saying concrete line I linked above has an issue - just saying we need to revisit all usages.

Reply:
Yes, you are right. I tend to maybe just make the default of numExpectedRows to 0. I am adding tests right now.

Reply:
I think defaulting it to `0` would bypass the intention here. `-1` should be read as UNKOWN, while `0` could be a valid real value (empty table).

Reply:
If it would not be a primitive it could be `null`. However, since we are doing mathematical operations on it `-1` is a  dangerous.

Reply:
What if don't expose `long numExpectedRows()` but rather `Consumer<Long> `in `LogicalPlan` - and default implementation would be smth like `if (numExpectedRows >= 0) consumer.accept `?

Reply:
Another simple idea could be an `isEmpty()` method on stats which could be used on the caller's side, to decide if stats should be skipped.

Reply:
It's common to use `-1` as a sentinel value to indicate a unknown state. Changing the type to `Integer` or `Long` would introduce boxing, and a second method like `isEmpty()` would add indirection without being any more safe as consumers would still have to know that they need to check it.

One question is if we actually need to distinguish between "no stats available" and "table has 0 records". If not I'd tend to go with 0 as default. But I'd think we should treat the two differently, and opt out of re-ordering if the stats are missing from one side.

Looks like there are other areas where this is currently handled incorrect. For example: https://github.com/crate/crate/blob/bbe7c7a28f409b35461dcd70051bb230fc065b88/server/src/main/java/io/crate/planner/operators/Union.java#L242. One could argue that if one side is -1, returning the count of the other side would be incorrect, and it would have to be -1. This kinda reinforces to me that we should treat "no stats available" as a separate case.